core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexUpgradeService.java

```diff
@@ -239,4 +261,67 @@ private IndexMetaData addDefaultUnitsIfNeeded(IndexMetaData indexMetaData) {
         // No changes:
         return indexMetaData;
     }
+
+
+    /**
+     * Checks the mappings of the mappings for compatibility with the current version
+     */
+    private IndexMetaData upgradeMappings(IndexMetaData indexMetaData) {
+        Index index = new Index(indexMetaData.getIndex());
+        Settings settings = indexMetaData.settings();
+        try {
+            SimilarityLookupService similarityLookupService = new SimilarityLookupService(index, settings);
+            // We cannot instantiate real analysis server at this point because the node might not have
+            // been started yet. However, we don't really need real analyzers at this stage - so we can fake it
+            try (AnalysisService analysisService = new FakeAnalysisService(index, settings)) {
+                try (MapperService mapperService = new MapperService(index, settings, analysisService, similarityLookupService, scriptService)) {
+                    for (ObjectCursor<MappingMetaData> cursor : indexMetaData.getMappings().values()) {
+                        MappingMetaData mappingMetaData = cursor.value;
+                        mapperService.merge(mappingMetaData.type(), mappingMetaData.source(), false, false);
+                    }
+                }
+            }
+        } catch (Exception ex) {
+            // Wrap the inner exception so we have the index name in the exception message
+            throw new IllegalStateException("unable to upgrade the index [" + indexMetaData.getIndex() + "], reason: [" + ex.getMessage() + "]", ex);
+        }
+        return indexMetaData;
+    }
+
+    /**
+     * Marks index as upgraded so we don't have to test it again
+     */
+    private IndexMetaData markAsUpgraded(IndexMetaData indexMetaData) {
+        Settings settings = Settings.builder().put(indexMetaData.settings()).put(IndexMetaData.SETTING_VERSION_UPGRADED, Version.CURRENT).build();
+        return IndexMetaData.builder(indexMetaData).settings(settings).build();
+    }
+
+    /**
+     * A fake analysis server that returns the same keyword analyzer for all requests
+     */
+    private static class FakeAnalysisService extends AnalysisService {
+
+        private Analyzer fakeAnalyzer = new Analyzer() {
+            @Override
+            protected TokenStreamComponents createComponents(String fieldName) {
+                throw new UnsupportedOperationException("shouldn't be here");
+            }
+        };
+
+        public FakeAnalysisService(Index index, Settings indexSettings) {
+            super(index, indexSettings);
+        }
+
+        @Override
+        public NamedAnalyzer analyzer(String name) {
+            return new NamedAnalyzer(name, fakeAnalyzer);
+        }
+
+        @Override
+        public void close() {
+            fakeAnalyzer.close();
```

Comment:
The parent class has a non-empty close method, so while you are right - there is nothing really happens there, I don't think it causes any harm to close it and it will make reasoning about the code easier.